### ISSUE TITLE
FormSplitter 'no-op' isn't actually a no-op

### ISSUE THREAD

**Issue by dorugeber** [firedrakeproject/firedrake#484]
The FormSplitter claims to... "*Split a form into a subtree for each component of the mixed space it is built on. This is a no-op on forms over non-mixed spaces.*"

Actually, it appears to split the form by `integral_type`, which isn't done otherwise.

This is clearly related to the various empty-integrand bug reports that we've had.  Not sure if this is 'just' a docstring bug or whether it's more pervasive.

Sample code:
```
from firedrake import *
mesh = UnitSquareMesh(5, 5)
P1 = FunctionSpace(mesh, "CG", 1)
u = TrialFunction(P1)
v = TestFunction(P1)
f = Function(P1)
f.assign(1)

a = inner(u, v)*dx
L = inner(f, v)*dx + inner(f, v)*ds(3)

u = Function(P1)
solve(a == L, u)
print u.dat.data
```

If one shells out in `compile_form` in `firedrake/ffc_interface.py`, after the line
`fd = compute_form_data(form)`,
then for the RHS, one gets the following output:

```
In [5]: print form
{ v_0 * function_9 } * dx(<Domain built from <triangle cell in 2D> with label None and coordinates Coefficient(VectorElement('Lagrange', Domain(Cell('triangle', 2), label=None, data='<data with id 139653886655824>'), 1, dim=2, quad_scheme=None), 0)>[everywhere], {})
  +  { -1 * v_0 * function_6 } * dx(<Domain built from <triangle cell in 2D> with label None and coordinates Coefficient(VectorElement('Lagrange', Domain(Cell('triangle', 2), label=None, data='<data with id 139653886655824>'), 1, dim=2, quad_scheme=None), 0)>[everywhere], {})
  +  { -1 * v_0 * function_6 } * ds(<Domain built from <triangle cell in 2D> with label None and coordinates Coefficient(VectorElement('Lagrange', Domain(Cell('triangle', 2), label=None, data='<data with id 139653886655824>'), 1, dim=2, quad_scheme=None), 0)>[3], {})

In [6]: bar = [foo[0][1] for foo in FormSplitter().split(form)]

In [7]: len(bar)
Out[7]: 2

In [8]: print bar[0]
{ v_0 * function_9 + -1 * v_0 * function_6 } * dx(<Domain built from <triangle cell in 2D> with label None and coordinates Coefficient(VectorElement('Lagrange', Domain(Cell('triangle', 2), label=None, data='<data with id 139653886655824>'), 1, dim=2, quad_scheme=None), 0)>[everywhere], {})

In [9]: print bar[1]
{ -1 * v_0 * function_6 } * ds(<Domain built from <triangle cell in 2D> with label None and coordinates Coefficient(VectorElement('Lagrange', Domain(Cell('triangle', 2), label=None, data='<data with id 139653886655824>'), 1, dim=2, quad_scheme=None), 0)>[3], {})
```

The "F == 0" version of this, which is the one that leads to bugs, is
```
...
v = TestFunction(P1)
sol = Function(P1)
f = Function(P1)
f.assign(1)

F = inner(sol, v)*dx - inner(f, v)*dx - inner(f, v)*ds(3)
solve(F == 0, sol)
```

```
In [1]: print form
{ d/dfj { v_0 * function_6 }, with fh=ExprList(*(function_6,)), dfh/dfj = ExprList(*(v_1,)), and coefficient derivatives ExprMapping(*()) } * dx(<Domain built from <triangle cell in 2D> with label None and coordinates Coefficient(VectorElement('Lagrange', Domain(Cell('triangle', 2), label=None, data='<data with id 140386065845648>'), 1, dim=2, quad_scheme=None), 0)>[everywhere], {})
  +  { d/dfj { -1 * v_0 * function_7 }, with fh=ExprList(*(function_6,)), dfh/dfj = ExprList(*(v_1,)), and coefficient derivatives ExprMapping(*()) } * dx(<Domain built from <triangle cell in 2D> with label None and coordinates Coefficient(VectorElement('Lagrange', Domain(Cell('triangle', 2), label=None, data='<data with id 140386065845648>'), 1, dim=2, quad_scheme=None), 0)>[everywhere], {})
  +  { d/dfj { -1 * v_0 * function_7 }, with fh=ExprList(*(function_6,)), dfh/dfj = ExprList(*(v_1,)), and coefficient derivatives ExprMapping(*()) } * ds(<Domain built from <triangle cell in 2D> with label None and coordinates Coefficient(VectorElement('Lagrange', Domain(Cell('triangle', 2), label=None, data='<data with id 140386065845648>'), 1, dim=2, quad_scheme=None), 0)>[3], {})

In [2]: bar = [foo[0][1] for foo in FormSplitter().split(form)]

In [3]: len(bar)
Out[3]: 2

In [4]: print bar[0]
{ (d/dfj { v_0 * function_6 }, with fh=ExprList(*(function_6,)), dfh/dfj = ExprList(*(v_1,)), and coefficient derivatives ExprMapping(*())) + (d/dfj { -1 * v_0 * function_7 }, with fh=ExprList(*(function_6,)), dfh/dfj = ExprList(*(v_1,)), and coefficient derivatives ExprMapping(*())) } * dx(<Domain built from <triangle cell in 2D> with label None and coordinates Coefficient(VectorElement('Lagrange', Domain(Cell('triangle', 2), label=None, data='<data with id 140386065845648>'), 1, dim=2, quad_scheme=None), 0)>[everywhere], {})

In [5]: print bar[1]
{ d/dfj { -1 * v_0 * function_7 }, with fh=ExprList(*(function_6,)), dfh/dfj = ExprList(*(v_1,)), and coefficient derivatives ExprMapping(*()) } * ds(<Domain built from <triangle cell in 2D> with label None and coordinates Coefficient(VectorElement('Lagrange', Domain(Cell('triangle', 2), label=None, data='<data with id 140386065845648>'), 1, dim=2, quad_scheme=None), 0)>[3], {})
```

**Comment by kynan** [comment]
You're absolutely right, it's definitely a docstring bug. Suppose the behaviour could be changed to return the original form if it's not mixed.

**Comment by wence-** [comment]
Fixed in #691
